radix-engine/src/kernel/kernel.rs

```diff
@@ -190,53 +220,51 @@ where
 
             // Run
             let output = M::invoke_upstream(args, self)?;
-
-            let mut update = CallFrameUpdate {
-                nodes_to_move: output.owned_node_ids().clone(),
-                node_refs_to_copy: output.references().clone(),
-            };
-
-            // Handle execution finish
-            M::on_execution_finish(&mut update, self)?;
+            let mut message = Message::from_indexed_scrypto_value(&output);
 
             // Auto-drop locks again in case module forgot to drop
             self.current_frame
                 .drop_all_locks(&mut self.heap, self.store)
                 .map_err(CallFrameError::UnlockSubstateError)
                 .map_err(KernelError::CallFrameError)?;
 
-            (output, update)
+            // Handle execution finish
+            M::on_execution_finish(&mut message, self)?;
+
+            (output, message)
         };
 
-        // Pop call frame
+        // Move
         {
-            let mut parent = self.prev_frame_stack.pop().unwrap();
+            let parent = self.prev_frame_stack.last_mut().unwrap();
```

Comment:
Is there a reason you didn't want to pop the parent yet?

Reply:
I think we shouldn't change the frame stack structure at the time of processing returns.

As you can tell, we have a dedicate stage for "pop call frame".